wasm/validation.cpp

```diff
@@ -0,0 +1,373 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "wasm/validation.h"
+#include "util/variant.h"
+#include "wasm/instructions.h"
+#include "wasm/types.h"
+#include "wasm/wasm.h"
+
+#include <tl/expected.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+#include <string_view>
+#include <variant>
+#include <vector>
+
+namespace wasm::validation {
+
+using namespace wasm;
+using namespace wasm::instructions;
+
+using ValueOrUnknown = std::variant<std::monostate, ValueType>;
+
+// NOLINTNEXTLINE(cert-err58-cpp)
+std::map<ValidationError, std::string_view> const validation_error_str = {
+        {ValidationError::ControlStackEmpty, "Attempted to pop from the control stack, but the control stack is empty"},
+        {ValidationError::LabelInvalid, "Attempted to branch to a label which isn't valid"},
+        {ValidationError::LocalUndefined, "Attempted to index a local which isn't defined in the current code entry"},
+        {ValidationError::MemoryBadAlignment, "Attempted a load or store with a bad alignment value"},
+        {ValidationError::MemoryEmpty, "Attempted a load, but memory is empty"},
+        {ValidationError::MemorySectionUndefined, "Attempted a load or store, but no memory section was defined"},
+        {ValidationError::TypeSectionUndefined,
+                "Encountered a control instruction with a type use, but no type section was defined"},
+        {ValidationError::TypeUndefined, "Attempted to index a type which isn't defined in the type section"},
+        {ValidationError::UnknownInstruction, "Unknown instruction encountered"},
+        {ValidationError::ValueStackHeightMismatch,
+                "Value stack height on exiting a control frame does not match the height on entry"},
+        {ValidationError::ValueStackUnderflow, "Attempted to pop from value stack, but stack height would underflow"},
+        {ValidationError::ValueStackUnexpected,
+                "Attempted to pop expected value from value stack, but got unexpected value"},
+};
+
+std::string_view to_string(ValidationError err) {
+    return validation_error_str.at(err);
+}
+
+struct ControlFrame {
+    Instruction i;
+    std::vector<ValueType> params;
+    std::vector<ValueType> results;
+    std::size_t stack_height;
+    bool unreachable;
+};
+
+namespace {
+
+void push_val(std::vector<ValueOrUnknown> &value_stack, ValueOrUnknown const &val) {
+    value_stack.push_back(val);
```

Comment:
I think it might be nicer to have something like:
```c++
struct Validator {
  std::vector<ValueOrUnknown> value_stack;
  std::vector<ControlFrame> control_stack;

  tl::expected<ValueOrUnknown, ValidationError> pop_val();
  ...
};
```

Having to pass these stacks to most functions feels like unnecessary boilerplate. What do you think?